Add accordion component

im36-123 opened the pull request:
I added 3 components for accordion.

im36-123 commented:
@nabeliwo @ouji-miyahara 
Please review !

im36-123 commented:
ah, this component name is AccordionPanel actually. I will fix it.

im36-123 commented:
Fixed !

nabeliwo commented on src/index.ts:
```diff
@@ -32,6 +32,7 @@ export { BlankImage } from './components/Image'
 export { Heading } from './components/Heading'
 export { HeadlineArea } from './components/HeadlineArea'
 export { Select } from './components/Select'
+export { Accordion, AccordionContent, AccordionTrigger } from './components/Accordion'
```
[MUST] Can't build due to error. `TS2307: Cannot find module './components/Accordion'.`

im36-123 replied:
I'm sorry for my mistakes.
Fixed this commit. -> https://github.com/kufu/smarthr-ui/pull/408/commits/1d0dd7516b2c234721876c0173e900dfb084b1d5

im36-123 commented:
@nabeliwo 

Thank you for your feedback, and discussion.
I improved my component like below.

## State management
State of expand changed to be managed inside components.

## expandableMultiply and defaultExpanded props
added 2 new props.
https://github.com/kufu/smarthr-ui/tree/add-accordion/src/components/AccordionPanel#props

## AccordionPanelItem
added AccordionPanelItem component, that contain AccordionPanelTrigger and AccordionPanelContent

im36-123 commented:
I will investigate animation issues in other PR

nabeliwo commented on src/components/AccordionPanel/AccordionPanelItem.tsx:
```diff
@@ -0,0 +1,25 @@
+        name,
+      }}
+    >
+      <div className={className}> {children}</div>
```
[ask] Why do you put a space in front of `{children}`?

im36-123 replied:
This is just typo

nabeliwo commented on src/components/AccordionPanel/AccordionPanelTrigger.tsx:
```diff
@@ -0,0 +1,111 @@
+
+  const handleClick = () => {
+    return onClick(name, !isExpanded)
+  }
```
[IMO] There is no need to wrap it with a function because it is not memoized.

nabeliwo commented on src/components/AccordionPanel/AccordionPanelTrigger.tsx:
```diff
@@ -0,0 +1,111 @@
+      id={`${name}-trigger`}
+      className={buttonClassNames}
+      aria-expanded={!!isExpanded}
+      aria-controls={`${name}-content`}
```
nice a11y 👍

nabeliwo commented on src/components/AccordionPanel/AccordionPanel.tsx:
```diff
@@ -0,0 +1,52 @@
+    (itemName: string, isExpanded: boolean) => {
+      if (expandableMultiply) {
+        isExpanded ? expandedItems.set(itemName, itemName) : expandedItems.delete(itemName)
+        setExpanded(new Map(expandedItems))
```
[MUST] Rewriting value directly for `expandedItems` state. Instead of creating a new Map when executing setExpanded, execute set or delete on the newly created Map from expandedItems.

nabeliwo commented on src/components/AccordionPanel/AccordionPanel.tsx:
```diff
@@ -0,0 +1,52 @@
+  // fires onClick after state of expanded changed
+  React.useEffect(() => {
+    if (onClick) onClick(mapToArray(expandedItems))
+  }, [expandedItems, onClick])
```
[ask] Can this onClick be passed to AccordionPanelTrigger as onClick?
I think that the expandedItems value at the time of click can also be derived on the AccordionPanelTrigger side.

nabeliwo commented on src/components/AccordionPanel/AccordionPanel.tsx:
```diff
@@ -0,0 +1,52 @@
+
+  return (
+    <AccordionPanelContext.Provider value={{ onClick: handleClick, expandedItems, icon }}>
+      <div className={className} children={children} />
```
[IMO] Please pass children in div instead of this format. I think it is pointed out by eslint.

nabeliwo commented on src/components/AccordionPanel/AccordionPanel.tsx:
```diff
@@ -0,0 +1,52 @@
+
+type Props = {
+  children: React.ReactNode
+  icon?: 'left' | 'right' | 'none'
```
[IMO] The name of the props `icon` is not intuitive. Whether or not the icon is displayed and the display position of the icon are different interests. It is difficult to guess 'left' | 'right' | 'none' from the name `icon`.
There should be boolean props to display the icon and 'left' | 'right' optional props to determine the position of the icon.

im36-123 commented on src/components/AccordionPanel/index.ts:
```diff
@@ -0,0 +1,3 @@
+export { AccordionPanelItem } from './AccordionPanelItem'
+export { AccordionPanelContent } from './AccordionPanelContent'
+export { AccordionPanelTrigger } from './AccordionPanelTrigger'
```
😳

im36-123 commented on src/components/AccordionPanel/README.md:
```diff
@@ -0,0 +1,36 @@
+| icon               | -        | **string** <br> left &#124; right &#124; none | left         | Position of icon. If sets 'none', icon is not display.                                           |
+| expandableMultiply | -        | **boolean**                                   | false        | If true, allow to expand accordion panel multiply.                                               |
+| defaultExpanded    | -        | **string[]**                                  | []           | Expands accordion with name matching value in array.                                             |
+| onClick            | ✓        | **function**                                  | -            | Fires when the AccordionPanelTrigger clicked. <br><br>function: (expandedList: string[]) => void |
```
Should I move `function: (expandedList: string []) => void` in the Description column to the type column?
Other documents (eg Button, CheckBox) have the same format as this PR.

im36-123 commented:
Thank you for your feedback !!
I fixed my code and left my response. Please check.

nabeliwo commented on src/libs/map.ts:
```diff
@@ -0,0 +1,9 @@
+export const getIsInclude = <K, V>(map: Map<K, V>, key: K) => !!map.get(key)
+
+export const mapToArray = <K, V>(map: Map<K, V>) => Array.from(map.keys())
```
[MUST] The actual behavior does not match the name `mapToArray`.
```ts
const map = new Map()

map.set('hoge', { name: 'hoge' })
map.set('fuga', { name: 'fuga' })

console.log(mapToArray(map))

/*
Expected value from the name 'mapToArray'
[
  ['hoge', { name: 'hoge' }],
  ['fuga', { name: 'fuga' }],
]

Actual value
['hoge', 'fuga']
*/
```

In the case of the current behavior, I think the name `mapToKeyArray` is appropriate.

im36-123 replied:
fixed: https://github.com/kufu/smarthr-ui/pull/408/commits/f75505e7e2c57f1b21169e3ebabc0411a3f69658

nabeliwo commented on src/libs/map.ts:
```diff
@@ -0,0 +1,9 @@
+
+export const mapToArray = <K, V>(map: Map<K, V>) => Array.from(map.keys())
+
+export const arrayToMap = <T>(array: T[]) => {
```
[MUST] Like `mapToKeyArray`, `arrayToMap` does not match the expected behavior from the name. I think the name `flatArrayToMap` is appropriate.

nabeliwo commented on src/components/AccordionPanel/AccordionPanelTrigger.tsx:
```diff
@@ -0,0 +1,114 @@
+
+  const handleClick = useCallback(() => {
+    onClickTrigger(name, !isExpanded)
+    if (onClickProps) onClickProps(mapToArray(expandedItems))
```
[MUST] If `expandedItems` is passed to `onClickProps` here, it will not match the expected behavior. An array with `name` of the clicked element added to `expandedItems` must be passed to `onClickProps`.

im36-123 commented on src/components/AccordionPanel/AccordionPanelItem.tsx:
```diff
@@ -0,0 +1,25 @@
+  className?: string
+}
+
+type ContextType = Omit<Props, 'children'>
```
fixed: https://github.com/kufu/smarthr-ui/pull/408/commits/74edb9a02493b61ca6e709007a550faaedb25c85

im36-123 commented on src/components/AccordionPanel/README.md:
```diff
@@ -0,0 +1,41 @@
+} from 'smarthr-ui'
+
+<AccordionPanel
+  icon="none"
```
fixed: https://github.com/kufu/smarthr-ui/pull/408/commits/b6b91be67330d7da4346663d2e7206d13a626bce

nabeliwo replied:
@im36-123 Please fix this also.
https://github.com/kufu/smarthr-ui/pull/408/files#diff-851b0528dd750e3a027eede30c200a07R33

im36-123 commented:
All fix is done. Please review again 🙏

im36-123 commented:
I changed the base branch feature/ttmz to master.

nabeliwo commented on src/components/AccordionPanel/README.md:
```diff
@@ -0,0 +1,42 @@
+
+| Name               | Required | Type                                          | DefaultValue | Description                                                                                      |
+| ------------------ | -------- | --------------------------------------------- | ------------ | ------------------------------------------------------------------------------------------------ |
+| icon               | -        | **string** <br> left &#124; right &#124; none | left         | Position of icon. If sets 'none', icon is not display.                                           |
```
https://github.com/kufu/smarthr-ui/pull/408#discussion_r345016511

im36-123 commented:
Thank you for your great feedbacks !!!
